index.html

```diff
@@ -3931,6 +3939,93 @@ <h2>Identification</h2>
+  canonicalization needs to select only one. 
+  </p><p>
+  Some of the choices leading to the efficient computation of a canonical form, 
+  such as the suppression of white space and sorting of elements in objects 
+  do make the representation less human readable.
+  In general, the canonical form presented here is optimized for machine-to-machine
+  use cases such as signing and to accommodate devices with limited network and
+  compute capabilities.
+  Compactness and ease of implementation
+  has been emphasized over human readability.
+  </p>
+  <p>Computing a canonical form directly from an RDF triple representation is
+  difficult so the starting point for the process described here is 
+  the JSON serialization of a TD.  
+  The following sequence of transformations are then applied to this serialization: 
+  <ol>
+  <li>Remove any elements that are given their default value. 
+  For example, the default value of <code>observable</code>
+  is <code>false</code>.  
+  If an input TD explictly assigns the value <code>false</code> to <code>observable</code>,
+  this element and the redundant value will be removed.</li>
+  <li>If removal of entries with default values results in duplicates in any <code>forms</code>
+  array, remove all but one duplicate in each such array.</li>
+  <li>Rewrite values that can be expressed as either an array or as a single value 
+  as a single value if there is only one element.  
+  In other words, remove square brackets around
+  arrays of single elements where this is permitted.
+  </li>
+  <li>For all extension vocabulary terms, sort the base URIs, assign system-generated prefixes 
+  using the template <code>e{n}</code> (where <code>{n}</code> is an integer sequence number starting
+  at 1), and use these prefixes in place of base URIs. Note: some systems discard the prefixes
+  and expand all vocabulary terms into URIs.  This rule allows the canonical form to be
+  generated even in this case. It does, however, lead to the loss of the original prefixes.</li>
+  <li>Apply the JSON Canonicalization Scheme [[!RFC8785]], which among other transformations,
+  sorts object members by name, removes white space, and normalizes number representations.</li>
+  <li>At the top level object of a TD, elements should be reordered as follows:
+    <code>@context</code>,
+    <code>@type</code>,
+    <code>id</code>,
+    <code>title</code>,
+    <code>titles</code>,
+    <code>description</code>,
+    <code>descriptions</code>,
+    <code>version</code>,
+    <code>created</code>,
+    <code>modified</code>,
+    <code>support</code>,
+    any extension terms (in sorted order),
+    <code>links</code>,
+    <code>securityDefinitions</code>,
+    <code>security</code>,
+    <code>forms</code>,
+    <code>properties</code>,
+    <code>actions</code>, and
+    <code>events</code>.
+  This reordering is to make single-pass processing easier, so that (for example) definitions 
+  occur prior to their first use.</li> 
```

Comment:
I have a feeling that this is mostly to make TDs human readable and better looking in the spec document.

Most importantly and as mentioned before, suggesting RFC8785 JSON Canonicalization Scheme for only parts of object and a different sorting order for the rest means implementations will not be able to use available libraries or any kind of existing JSON serializer for that matter. Note that in most programming languages, JSON is purely the serialized output of an internal object (class/struct instance) or dictionary/map. For generic directories which need to blindly host and give back TDs with extensions (with and without namespace prefixes), anything other than an alphanumeric sorting will require complex and inefficient custom serialization.

I understand that single-pass processing may be a consumer requirement, but the suggested ordering doesn't match every use case. Imagine the following example from a smart city application:
```json
{
    "@context": [
        "https://www.w3.org/2019/wot/td/v1",
        "<domain-specific-context-url>"
    ],
    "id": "urn:uuid:15eb6579-29b9-4329-a952-3a679eab8cc4",
    "type": "motionSensor", // alias for @type
    "site": "alpha", // extension
    "projectID": "80581098-5f7f-4e8b-be02-d1d82fb4b161", // extension
    "ifcZoneID": "ea971d5e-96df-4369-9ee5-7a5bac67437d", // extension
    "ifcSpaceID": "a7b3931f-aab5-4e4f-88de-ad369a2c4393", // extension
    ...
    "properties": {
       ...
     },
     ...
}
```
Here, the title(s) and description(s) have no use in single-processing. Instead, several domain-specific extensions need to be consumed to know the location of the thing before processing the rest.

Thus, I suggest not prescribing any custom ordering for canonicalization and relying fully on RFC8785. The servers which have single-processing consumers can provide appropriate custom ordering (which we cannot generalize) through server-driven content negotiation.

Reply:
After further thought I am inclined to agree with you.  There are enough other things that make canonical form non-human-readable that it is moot.  Consumers will have to simply have enough memory to read an entire TD into memory before processing it.